### ISSUE TITLE
Unable to install package with custom pypi source

### ISSUE THREAD

**Issue by ocavue** [python-poetry/poetry#1841]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: macOS 10.14.6 / docker image python:3.5
- **Poetry version**: 1.0.0
- **Link of a ~~[Gist](https://gist.github.com/)~~ repository with the contents of your pyproject.toml file**: https://github.com/ocavue/poetry_issue_celery

## Issue

The steps to reproduce this bug is in the [readme](https://github.com/ocavue/poetry_issue_celery).

In short, I can't install any package with a custom pypi source (https://pypi.douban.com/simple/). `pip3 install celery -i https://pypi.douban.com/simple/` works find, so this should be a bug of poetry.

**Comment by TheErk** [comment]
https://pypi.douban.com seems to resolve to https://pypi.doubanio.com/ may be poetry does not follow redirection?

I do personnally use custom pypi source and in my use case it works as expected.

**Comment by ocavue** [comment]
@TheErk Thanks for your comment.

https://pypi.douban.com/ return a 301 to https://pypi.doubanio.com/. After using https://pypi.doubanio.com/ in pyproject.toml, I can install packages.

So the problem here is that poetry doesn't follow the 301 redirect to find package index.

**Comment by ocavue** [comment]
I did some little research. poetry use `requests.get()` to fetch packages from PyPI repository. `request.get()` will follow 301 redirection by [default](https://requests.readthedocs.io/en/v2.9.1/user/quickstart/#redirection-and-history). The reason that poetry can't install packages from https://pypi.douban.com/ is that Douban server return *418 I'm a teapot*. I guess that this is an anti-scraping logic.


Use `curl` to visit pypi.douban.com. Get 301:


```
$ curl -v 'https://pypi.douban.com/simple/celery/'
...
< HTTP/1.1 301 Moved Permanently
...
```


Use `requests` to visit pypi.douban.com. Get 418:


```
In [1]: import requests

In [2]: requests.get('https://pypi.douban.com/simple/celery/')
Out[2]: <Response [418]>

In [3]:  requests.get('https://pypi.doubanio.com/simple/celery/')
Out[3]: <Response [200]>

In [4]: requests.get('https://pypi.douban.com/simple/celery/', headers={'user-agent': 'poetry'})
Out[4]: <Response [200]>
```


To avoid this `418` response, we can simply pass a `user-agent` into the request header.


Since adding a header in the request can fix this issue and I couldn't see any disadvantage about this change, I can submit a PR to do it if you agree with me.  @sdispater 

---


[Douban](https://en.wikipedia.org/wiki/Douban) is one of the biggest Chinese companies that use Python as their main back-end language. The PyPI mirror their offer is very famous in the Chinese Python developer community.

**Comment by thomaszdxsn** [comment]
better to have log output for source url on `poetry install -v`

**Comment by ocavue** [comment]
@thomaszdxsn 

```
$ poetry install -v
Using virtualenv: /Users/ocavue/code/github/poetry_issue_celery/.venv
Updating dependencies
Resolving dependencies... (0.2s)

[SolverProblemError]
Because poetry-issue-celery depends on celery (*) which doesn't match any versions, version solving failed.

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/clikit/console_application.py", line 131, in run
    status_code = command.handle(parsed_args, io)
  File "/usr/local/lib/python3.7/site-packages/clikit/api/command/command.py", line 120, in handle
    status_code = self._do_handle(args, io)
  File "/usr/local/lib/python3.7/site-packages/clikit/api/command/command.py", line 171, in _do_handle
    return getattr(handler, handler_method)(args, io, self)
  File "/usr/local/lib/python3.7/site-packages/cleo/commands/command.py", line 92, in wrap_handle
    return self.handle()
  File "/usr/local/lib/python3.7/site-packages/poetry/console/commands/install.py", line 63, in handle
    return_code = installer.run()
  File "/usr/local/lib/python3.7/site-packages/poetry/installation/installer.py", line 74, in run
    self._do_install(local_repo)
  File "/usr/local/lib/python3.7/site-packages/poetry/installation/installer.py", line 161, in _do_install
    ops = solver.solve(use_latest=self._whitelist)
  File "/usr/local/lib/python3.7/site-packages/poetry/puzzle/solver.py", line 36, in solve
    packages, depths = self._solve(use_latest=use_latest)
  File "/usr/local/lib/python3.7/site-packages/poetry/puzzle/solver.py", line 190, in _solve
    raise SolverProblemError(e)

```

**Comment by thomaszdxsn** [comment]
> @thomaszdxsn
> 
> ```
> $ poetry install -v
> Using virtualenv: /Users/ocavue/code/github/poetry_issue_celery/.venv
> Updating dependencies
> Resolving dependencies... (0.2s)
> 
> [SolverProblemError]
> Because poetry-issue-celery depends on celery (*) which doesn't match any versions, version solving failed.
> 
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.7/site-packages/clikit/console_application.py", line 131, in run
>     status_code = command.handle(parsed_args, io)
>   File "/usr/local/lib/python3.7/site-packages/clikit/api/command/command.py", line 120, in handle
>     status_code = self._do_handle(args, io)
>   File "/usr/local/lib/python3.7/site-packages/clikit/api/command/command.py", line 171, in _do_handle
>     return getattr(handler, handler_method)(args, io, self)
>   File "/usr/local/lib/python3.7/site-packages/cleo/commands/command.py", line 92, in wrap_handle
>     return self.handle()
>   File "/usr/local/lib/python3.7/site-packages/poetry/console/commands/install.py", line 63, in handle
>     return_code = installer.run()
>   File "/usr/local/lib/python3.7/site-packages/poetry/installation/installer.py", line 74, in run
>     self._do_install(local_repo)
>   File "/usr/local/lib/python3.7/site-packages/poetry/installation/installer.py", line 161, in _do_install
>     ops = solver.solve(use_latest=self._whitelist)
>   File "/usr/local/lib/python3.7/site-packages/poetry/puzzle/solver.py", line 36, in solve
>     packages, depths = self._solve(use_latest=use_latest)
>   File "/usr/local/lib/python3.7/site-packages/poetry/puzzle/solver.py", line 190, in _solve
>     raise SolverProblemError(e)
> ```

我的意思是最好在--verbose状态下输出下载使用的pypi，我一直怀疑没有用到douban mirror，但是一直没有确定

**Comment by ocavue** [comment]
@thomaszdxsn

I have already found the reason of this issue as my comments above. We need not more discussion about the reason.

If you just want to avoid this issue, your can use https://mirrors.aliyun.com/pypi/simple.

**Comment by OnePieceLv** [comment]
the same issue to me.

**Comment by sinoroc** [comment]
@ocavue @OnePieceLv There is a PR that could help solve this issue. Would you be able to test it, give feedback? https://github.com/python-poetry/poetry/pull/3624

**Comment by ocavue** [comment]
@sinoroc Since both this issue and some other pull requests I submitted got zero response from the project maintainer, I have given up using this poetry and I will not contribute to this project anymore. Sorry for the inconvenience.

**Comment by sinoroc** [comment]
@ocavue I know, it can get frustrating. I am sure maintainers do their best, but it might still feel to far in between, I get it. That is exactly why here I was trying to get non-maintainers to collaborate with each other and get this as close to the finish line as possible to make the maintainers' work as easy as possible. Anyway, no worries, I understand. :)

**Comment by neersighted** [comment]
Closing this as stale / nothing to be done. Poetry can't follow a redirect when the server is actively sabotaging us by returning nonsense responses.
